src/messaging/ExchangeContext.h

```diff
@@ -0,0 +1,271 @@
+     *
+     *  @param[in]    sendFlags     Flags set by the application for the CHIP message being sent.
+     *
+     *  @param[in]    msgCtxt       A pointer to an application-specific context object to be associated
+     *                              with the message being sent.
+
+     *  @retval  #CHIP_ERROR_INVALID_ARGUMENT               if an invalid argument was passed to this SendMessage API.
+     *  @retval  #CHIP_ERROR_SEND_THROTTLED                 if this exchange context has been throttled when using the
+     *                                                       CHIP reliable messaging protocol.
+     *  @retval  #CHIP_ERROR_WRONG_MSG_VERSION_FOR_EXCHANGE if there is a mismatch in the specific send operation and the
+     *                                                       CHIP message protocol version that is supported. For example,
+     *                                                       this error would be generated if CHIP Reliable Messaging
+     *                                                       semantics are being attempted when the CHIP message protocol
+     *                                                       version is V1.
+     *  @retval  #CHIP_ERROR_NOT_CONNECTED                  if the context was associated with a connection that is now
+     *                                                       closed.
+     *  @retval  #CHIP_ERROR_INCORRECT_STATE                if the state of the exchange context is incorrect.
+     *  @retval  #CHIP_NO_ERROR                             if the CHIP layer successfully sent the message down to the
+     *                                                       network layer.
+     */
+    CHIP_ERROR SendMessage(uint32_t protocolId, uint8_t msgType, System::PacketBuffer * msgPayload, uint16_t sendFlags = 0,
+                           void * msgCtxt = nullptr);
+
+    /**
+     *  Tear down the Trickle retransmission mechanism by canceling the periodic timers
+     *  within Trickle and freeing the message buffer holding the Weave
+     *  message.
+     */
+    void TeardownTrickleRetransmit();
+
+    /**
+     *  Cancel the Trickle retransmission mechanism.
+     */
+    void CancelRetrans();
+
+    /**
+     *  Set delegate of the current ExchangeContext instance.
+     */
+    void SetDelegate(ExchangeContextDelegate * delegate) { mDelegate = delegate; }
+
+    /**
+     *  Return the delegate pointer of the current ExchangeContext instance.
+     */
+    ExchangeContextDelegate * GetDelegate() { return mDelegate; }
+
+    /*
+     * In order to use reference counting (see refCount below)
+     * we use a hold/free paradigm where users of the exchange
+     * can hold onto it while it's out of their direct control
+     * to make sure it isn't closed before everyone's ready.
+     */
+    void AddRef();
+    void Close();
+    void Abort();
+    void Release();
+    CHIP_ERROR StartTimerT();
```

Comment:
what is the timerT?

This seems to require doccomment. Also if this relates to trickle, maybe we can remove from initial PR.

Reply:
Added, I don't trickle feature is part of CRMP, it is not within the WRMP feature flag within wml, and managed by protocols. My understanding is CRMP is invisible to protocols.